src/rest_vol_dataset.c

```diff
@@ -1067,13 +1121,84 @@ RV_dataset_write(size_t count, void *dset[], hid_t mem_type_id[], hid_t _mem_spa
+        if (is_compound_subset) {
+            char     *cmpd_query_ptr   = cmpd_query;
+            ptrdiff_t cmpd_query_len   = 0;
+            size_t    num_cmpd_members = 0;
+
+            /* Retrieve compound information for request */
+            if (RV_get_cmpd_subset_nmembers(transfer_info[i].mem_type_id,
+                                            transfer_info[i].u.write_info.dense_cmpd_subset_dtype_id,
+                                            &num_cmpd_members) < 0)
+                FUNC_GOTO_ERROR(H5E_DATASET, H5E_CANTGET, FAIL, "can't get compound subset information");
+
+            /* Construct compound member selection string for request URL */
+            if (!(SERVER_VERSION_SUPPORTS_MEMBER_SELECTION(
+                    transfer_info[i].dataset->domain->u.file.server_info.version)))
+                FUNC_GOTO_ERROR(H5E_DATASET, H5E_UNSUPPORTED, FAIL,
+                                "compound member selection on write requires server >= 0.8.5");
+
+            /* If using element selection and member selection, preface member query with '&' to join it to
+               previous queries. Otherwise, it's the first query, so preface it with '?' */
+            if (has_selection_body) {
+                cmpd_query_ptr[0] = '&';
+                cmpd_query_ptr++;
+            }
+            else {
+                cmpd_query_ptr[0] = '?';
+                cmpd_query_ptr++;
+            }
+
+            memcpy(cmpd_query_ptr, COMPOUND_MEMBER_QUERY, strlen(COMPOUND_MEMBER_QUERY));
+            cmpd_query_ptr += strlen(COMPOUND_MEMBER_QUERY);
+
+            for (size_t j = 0; j < num_cmpd_members; j++) {
+                if ((member_name = H5Tget_member_name(transfer_info[i].mem_type_id, (unsigned)j)) == NULL)
+                    FUNC_GOTO_ERROR(H5E_DATASET, H5E_CANTGET, FAIL, "can't get name of compound member");
+
+                cmpd_query_len = cmpd_query_ptr - cmpd_query;
+
+                /*  URL encode the member name */
+                if ((url_encoded_member_name =
+                         curl_easy_escape(curl, member_name, (int)strlen(member_name))) == NULL)
+                    FUNC_GOTO_ERROR(H5E_DATASET, H5E_CANTENCODE, FAIL,
+                                    "can't URL-encode compound member name");
+
+                if ((size_t)cmpd_query_len + strlen(url_encoded_member_name) + 1 > URL_MAX_LENGTH)
+                    FUNC_GOTO_ERROR(H5E_DATASET, H5E_BADVALUE, FAIL,
+                                    "compound member names too long for URL");
+
+                /* Copy name to query string without null byte */
+                memcpy(cmpd_query_ptr, member_name, strlen(member_name));
+                cmpd_query_ptr += strlen(member_name);
+
+                /* If another member name will follow, add sepator */
+                if (j < num_cmpd_members - 1) {
+                    *cmpd_query_ptr = COMPOUND_MEMBER_SEPARATOR;
+                    cmpd_query_ptr++;
```

Comment:
It's likely worth adding another check for going past `URL_MAX_LENGTH` here before adding the separator and below before adding the NUL terminator just to be better assured that we aren't running off the buffer end.